### PR TITLE
Cache icons in PhysicalWebcollection

### PR DESCRIPTION
**Pull request by nishadg246**
Writes icons to cache along with PwsResult data instead
of fetching them each time on resume

### CODE REVIEW

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/UrlDeviceDiscoveryService.java [hunk 1]
```diff
@@ -194,7 +194,7 @@ public void onCreate() {
     super.onCreate();
     initialize();
     restoreCache();
-
+    triggerCallback();
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/UrlDeviceDiscoveryService.java]: There are no listeners at this point.  Why do you need this line?

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/UrlDeviceDiscoveryService.java [hunk 2]
```diff
@@ -211,6 +211,7 @@ public IBinder onBind(Intent intent) {
 
   @Override
   public boolean onUnbind(Intent intent) {
+    triggerCallback();
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/UrlDeviceDiscoveryService.java]: This shouldn't be needed...any update that has come through should already be communicated.

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 3]
```diff
@@ -106,7 +110,7 @@ public UrlDevice getUrlDeviceById(String id) {
    * Fetches cached URL metadata using the URL broadcasted by the Physical Web device.
    * @param broadcastUrl The URL broadcasted by the device.
    * @return Cached metadata relevant to the given URL.
-   */
+ */
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: fix

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 4]
```diff
@@ -127,10 +131,20 @@ public JSONObject jsonSerialize() {
 
     // Serialize the URL metadata
     JSONArray metadata = new JSONArray();
+    JSONArray iconMap = new JSONArray();
     for (PwsResult pwsResult : mBroadcastUrlToPwsResultMap.values()) {
       metadata.put(pwsResult.jsonSerialize());
+      JSONObject icon = new JSONObject();
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: Why not just a string?

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 5]
```diff
@@ -127,10 +131,20 @@ public JSONObject jsonSerialize() {
 
     // Serialize the URL metadata
     JSONArray metadata = new JSONArray();
+    JSONArray iconMap = new JSONArray();
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: Sholudn't this be an object?

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 6]
```diff
@@ -127,10 +131,20 @@ public JSONObject jsonSerialize() {
...
     for (PwsResult pwsResult : mBroadcastUrlToPwsResultMap.values()) {
       metadata.put(pwsResult.jsonSerialize());
+      JSONObject icon = new JSONObject();
+      if (getIcon(pwsResult.getIconUrl()) == null){
+        icon.put(ICON_KEY, new byte[0]);
+      } else {
+        icon.put(ICON_KEY, new String(Base64.encodeBase64(getIcon(pwsResult.getIconUrl()))));
+        System.out.println("icons in:" + getIcon(pwsResult.getIconUrl()).length);
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: Remove

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 7]
```diff
@@ -163,12 +177,18 @@ public static PhysicalWebCollection jsonDeserialize(JSONObject jsonObject)
 
     // Deserialize the URL metadata
     JSONArray metadata = jsonObject.getJSONArray(METADATA_KEY);
+    JSONArray icons = jsonObject.getJSONArray(ICON_MAP_KEY);
     for (int i = 0; i < metadata.length(); i++) {
       JSONObject pwsResultJson = metadata.getJSONObject(i);
       PwsResult pwsResult = PwsResult.jsonDeserialize(pwsResultJson);
+      JSONObject icon = icons.getJSONObject(i);
+      byte[] iconDecode = Base64.decodeBase64(icon.getString(ICON_KEY).getBytes());
+      System.out.println("icons out:" + iconDecode.length);
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: Remove

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 8]
```diff
@@ -35,6 +38,8 @@
   private static final String SCHEMA_VERSION_KEY = "schema";
   private static final String DEVICES_KEY = "devices";
   private static final String METADATA_KEY = "metadata";
+  private static final String ICON_MAP_KEY = "iconmap";
+  private static final String ICON_KEY = "icon";
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: not needed

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 9]
```diff
@@ -169,6 +180,13 @@ public static PhysicalWebCollection jsonDeserialize(JSONObject jsonObject)
       collection.addMetadata(pwsResult);
     }
 
+    JSONObject iconMap = jsonObject.getJSONObject(ICON_MAP_KEY);
+    Iterator<String> iconUrls = iconMap.keys();
+    while (iconUrls.hasNext()) {
+      String iconUrl = iconUrls.next();
+      collection.addIcon(iconUrl,Base64.decodeBase64(iconMap.getString(iconUrl).getBytes()));
+    }
+    System.out.println("icons: Deserialize");
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: Remove

File: java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java [hunk 10]
```diff
@@ -169,6 +180,13 @@ public static PhysicalWebCollection jsonDeserialize(JSONObject jsonObject)
       collection.addMetadata(pwsResult);
     }
 
+    JSONObject iconMap = jsonObject.getJSONObject(ICON_MAP_KEY);
+    Iterator<String> iconUrls = iconMap.keys();
```
Comment (cco3) [inline java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java]: Would you mind using a for loop here?

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/UrlDeviceDiscoveryService.java [hunk 11]
```diff
@@ -257,11 +256,8 @@ public void onDestroy() {
 
   @Override
   public void onUrlDeviceDiscovered(UrlDevice urlDevice) {
-    // check if device has already been discovered
-    if (mPwCollection.addUrlDevice(urlDevice)) {
-      return;
-    }
-    // if we have found a new device
+    // add Device and fetch results
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/UrlDeviceDiscoveryService.java]: Please add a comment on why we don't want to short circuit here (favicon might not be fetched).

**Comment by nishadg246** [comment]
@cco3

**Pull request merged**